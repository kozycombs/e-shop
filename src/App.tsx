import { Suspense, lazy } from "react";
import { Routes, Route } from "react-router-dom";
import Layout from "./components/layout/Layout";
import { APP_URL } from "./constants";
import Spinner from "./components/spinner/Spinner";

function App() {
  const ProductsPage = lazy(() => import("./scenes/products/Products"));
  const CartPage = lazy(() => import("./scenes/cart/Cart"));

  return (
    <Layout>
      <Suspense fallback={<Spinner />}>
        <Routes>
          <Route path={APP_URL.HOME} element={<ProductsPage />} />
          <Route path={APP_URL.CART} element={<CartPage />} />
          <Route path="*" element={<ProductsPage />} />
        </Routes>
      </Suspense>
    </Layout>
  );
}

export default App;
