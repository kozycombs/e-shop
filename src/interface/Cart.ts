export interface CartProduct {
  productId: number;
  quantity: number;
}

export interface Cart {
  id: number;
  userId: number;
  date: string;
  products: CartProduct[];
}

export interface CartResponse {
  id: number;
}
