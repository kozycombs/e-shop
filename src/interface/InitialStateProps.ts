export interface InitialStateProps<T> {
  data: T;
  error: string;
}
