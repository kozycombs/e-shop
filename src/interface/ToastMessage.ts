export interface ToastMessage {
  message: string;
  severity: "success" | "error";
}
