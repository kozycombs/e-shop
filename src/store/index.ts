import { configureStore } from "@reduxjs/toolkit";
import productReducer from "store/productSlice";
import cartReducer from "store/cartSlice";

export const store = configureStore({
  reducer: {
    products: productReducer,
    cart: cartReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;
