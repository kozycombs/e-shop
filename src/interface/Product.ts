import { Rating } from "interface/Rating";

export interface Product {
  id: number;
  title: string;
  price: number;
  description: string;
  category: string;
  image: string;
  rating: Rating;
}
