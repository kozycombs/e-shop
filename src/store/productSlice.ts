import { createSlice } from "@reduxjs/toolkit";
import { RootState } from ".";
import { Product } from "../interface/Product";

export const initialState: Product[] = [];

export const productSlice = createSlice({
  name: "products",
  initialState,
  reducers: {
    updateProducts: (state, action) => {
      state = action.payload;
    },
  },
});

export const { updateProducts } = productSlice.actions;

export default productSlice.reducer;

export const productsSelector = (state: RootState) => state.products;
