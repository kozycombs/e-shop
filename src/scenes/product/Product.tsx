import { FC, useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import {
  Alert,
  Box,
  Button,
  Divider,
  FormControl,
  Grid,
  MenuItem,
  Rating,
  Select,
  SelectChangeEvent,
  Snackbar,
  Typography,
} from "@mui/material";
import { fetchProduct, productsSelector } from "store/productSlice";
import { Product as ProductInterface } from "interface/Product";
import Spinner from "components/spinner/Spinner";
import { addToCart, cartSelector, updateCartProduct } from "store/cartSlice";
import { AppDispatch } from "store/index";
import { ToastMessage } from "interface/ToastMessage";
import { PRODUCT_QUATITY_COUNT } from "constants/index";
import ErrorBanner from "components/errorBanner/ErrorBanner";

const Product: FC = () => {
  const { id } = useParams();
  const dispatch = useDispatch<AppDispatch>();
  const { data: products } = useSelector(productsSelector);
  const { data: cart } = useSelector(cartSelector);
  const [loading, setLoading] = useState(false);
  const [product, setProduct] = useState<ProductInterface>();
  const [quantity, setQuantity] = useState(1);
  const [toastMessage, setToastMessage] = useState<ToastMessage>({
    message: "",
    severity: "success",
  });
  const [error, setError] = useState("");
  const userId = 1; // This is a dummy user id, since there is no logged in functionality.
  const vertical = "top";
  const horizontal = "right";

  useEffect(() => {
    const getProduct = async () => {
      try {
        const result = await dispatch(fetchProduct(Number(id))).unwrap();
        setProduct(result);
        setLoading(false);
      } catch (error) {
        setError("Failed to load our product. Please try again later");
        setLoading(false);
      }
    };

    // If product list is empty get product detail from API
    // otherwiser get product detail from cached product list.

    if (products.length <= 0) {
      setLoading(true);
      getProduct();
    } else {
      setLoading(true);
      const product =
        products.find((currentProduct) => currentProduct.id === Number(id)) ||
        null;
      if (product) {
        setProduct(product);
      }
      setLoading(false);
    }
  }, [id, products, dispatch]);

  const resetToastMessage = () => {
    setToastMessage({ message: "", severity: "success" });
  };

  const handleQuantityChange = (event: SelectChangeEvent) => {
    setQuantity(Number(event.target.value));
  };

  const handleAddToCart = async () => {
    if (!product) {
      return;
    }

    try {
      setLoading(true);
      const date = new Date().toISOString();

      // If cart is null add to cart else update with cart id.
      if (!cart) {
        const result = await dispatch(
          addToCart({
            userId,
            date,
            products: [{ productId: product?.id, quantity }],
          })
        ).unwrap();
        if (result.id) {
          setToastMessage({
            message: "Product added to cart successfully",
            severity: "success",
          });
        } else {
          setToastMessage({
            message: "Failed to add product to cart",
            severity: "error",
          });
        }
      } else {
        const productFoundIndex = cart.products.findIndex(
          (currentProduct) => currentProduct.productId === product.id
        );

        if (productFoundIndex > -1) {
          const updatedProduct = {
            ...cart.products[productFoundIndex],
            quantity: cart.products[productFoundIndex].quantity + quantity,
          };
          const otherProducts = cart.products.filter(
            (currentProduct) => currentProduct.productId !== product.id
          );
          const products = [...otherProducts, updatedProduct];
          const result = await dispatch(
            updateCartProduct({ cartId: cart.id, userId, date, products })
          ).unwrap();
          if (result.id) {
            setToastMessage({
              message: "Product added to cart successfully",
              severity: "success",
            });
            setQuantity(1);
          } else {
            setToastMessage({
              message: "Failed to add product to cart",
              severity: "error",
            });
          }
        } else {
          const products = [
            ...cart.products,
            { productId: product?.id, quantity },
          ];
          const result = await dispatch(
            updateCartProduct({ cartId: cart.id, userId, date, products })
          ).unwrap();
          if (result.id) {
            setToastMessage({
              message: "Product added to cart successfully",
              severity: "success",
            });
            setQuantity(1);
          } else {
            setToastMessage({
              message: "Failed to add product to cart",
              severity: "error",
            });
          }
        }
      }
      setLoading(false);
    } catch (error) {
      setError("Failed to add product to cart. Please try again later");
      setLoading(false);
    }
  };

  return (
    <Box sx={{ pt: 4 }}>
      {loading && (
        <Box className="flex justify-center">
          <Spinner />
        </Box>
      )}

      {error.length > 0 && <ErrorBanner message={error} />}

      <Snackbar
        anchorOrigin={{ vertical, horizontal }}
        open={toastMessage?.message.length > 0}
        autoHideDuration={2000}
        onClose={resetToastMessage}
        key={vertical + horizontal}
      >
        <Alert
          onClose={resetToastMessage}
          severity={toastMessage?.severity}
          variant="filled"
          sx={{ width: "100%" }}
        >
          {toastMessage?.message}
        </Alert>
      </Snackbar>

      <Grid container spacing={{ xs: 4, md: 3 }} columns={{ xs: 1, sm: 4 }}>
        <Grid item xs={1} sm={2} sx={{ marginBottom: 10 }}>
          <Box
            component="img"
            sx={{
              height: "auto",
              width: "100%",
              maxHeight: { xs: 400 },
              objectFit: "contain",
            }}
            alt={product?.title}
            src={product?.image}
          />
        </Grid>
        <Grid item xs={1} sm={2}>
          <Box></Box>
          <Typography
            variant="h1"
            sx={{ mb: 1, fontWeight: "400", fontSize: 24 }}
          >
            {product?.title}
          </Typography>
          <Box sx={{ display: "flex", alignItems: "end", mb: 2 }}>
            <Typography
              component="span"
              sx={{
                mr: 1,
                fontSize: 16,
              }}
            >
              {product?.rating.rate}
            </Typography>
            <Rating
              name="read-only"
              value={Number(product?.rating.rate)}
              readOnly
            />
            <Typography
              component="span"
              sx={{
                ml: 1,
                mt: 1,
                fontSize: 16,
                textDecoration: "underline",
              }}
            >
              {product?.rating.count} reviews
            </Typography>
          </Box>
          <Typography
            component="p"
            sx={{ fontSize: 24, fontWeight: 400, mb: 4 }}
          >
            &pound;{Number(product?.price).toFixed(2)}
          </Typography>
          <Box sx={{ display: "flex", alignItems: "center", mb: 4 }}>
            <FormControl sx={{ m: 1, minWidth: 120 }} size="small">
              <Select
                id="quantity"
                data-test="quantity"
                value={`${quantity}`}
                onChange={handleQuantityChange}
              >
                {Array(PRODUCT_QUATITY_COUNT)
                  .fill("_")
                  .map((_, index) => (
                    <MenuItem key={index} value={index + 1}>
                      {index + 1}
                    </MenuItem>
                  ))}
              </Select>
            </FormControl>
            <Button
              variant="contained"
              size="medium"
              disableElevation
              onClick={handleAddToCart}
              sx={{ textTransform: "none" }}
            >
              Add to Cart
            </Button>
          </Box>
          <Divider />
          <Typography
            component="p"
            sx={{
              mb: 2,
              mt: 4,
              fontSize: 16,
            }}
          >
            {product?.description}
          </Typography>
        </Grid>
      </Grid>
    </Box>
  );
};

export default Product;
