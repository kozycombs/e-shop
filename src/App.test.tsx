import { act } from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { updateProducts } from "store/productSlice";
import { store } from "store";
import { Product } from "interface/Product";
import App from "App";

describe("<App />", () => {
  beforeAll(() => {
    const product: Product = {
      id: 16,
      title:
        "Lock and Love Women's Removable Hooded Faux Leather Moto Biker Jacket",
      price: 29.95,
      description:
        "100% POLYURETHANE(shell) 100% POLYESTER(lining) 75% POLYESTER 25% COTTON (SWEATER), Faux leather material for style and comfort / 2 pockets of front, 2-For-One Hooded denim style faux leather jacket, Button detail on waist / Detail stitching at sides, HAND WASH ONLY / DO NOT BLEACH / LINE DRY / DO NOT IRON",
      category: "women's clothing",
      image: "https://fakestoreapi.com/img/81XH0e8fefL._AC_UY879_.jpg",
      rating: {
        rate: 2.9,
        count: 340,
      },
    };

    store.dispatch(updateProducts([product]));
  });

  test("renders a product title", async () => {
    let component;
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(async () => {
      component = render(
        <Provider store={store}>
          <BrowserRouter>
            <App />
          </BrowserRouter>
        </Provider>
      );
    });

    expect(component).toMatchSnapshot();
  });
});
