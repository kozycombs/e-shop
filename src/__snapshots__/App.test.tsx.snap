// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders a product title 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1622sjm-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiContainer-root MuiContainer-maxWidthLg css-1j0emvn-MuiContainer-root"
        >
          <header
            class="flex justify-between items-center py-4"
          >
            <a
              class="font-semibold text-2xl text-slate-800"
              href="/"
            >
              E-Shop
            </a>
            <nav
              class="hidden sm:block"
            >
              <ul
                class="flex gap-8"
              >
                <li>
                  <a
                    class="text-slate-800 font-medium"
                    href="/"
                  >
                    All Products
                  </a>
                </li>
                <span
                  class="MuiBadge-root css-1c32n2y-MuiBadge-root"
                >
                  <li>
                    <a
                      class="text-slate-800 font-medium"
                      href="/cart"
                    >
                      Cart
                    </a>
                  </li>
                  <span
                    class="MuiBadge-badge MuiBadge-standard MuiBadge-invisible MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorPrimary css-y2h6q0-MuiBadge-badge"
                  />
                </span>
              </ul>
            </nav>
            <button
              aria-label="open drawer"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium bg-slate-800 css-ddp6wk-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="MenuIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </header>
        </div>
      </nav>
      <div
        class="MuiBox-root css-i9gxme"
      >
        <div
          class="MuiContainer-root MuiContainer-maxWidthLg css-1j0emvn-MuiContainer-root"
        >
          <main
            class="pt-[89px]"
          >
            <div
              class="MuiBox-root css-0"
            >
              <h3
                class="MuiTypography-root MuiTypography-h3 css-a5qzwy-MuiTypography-root"
              >
                Products
              </h3>
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-spacing-md-3 css-1bzsu2l-MuiGrid-root"
              >
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 css-1xucbe2-MuiGrid-root"
                >
                  <a
                    href="/product/16"
                  >
                    <div
                      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
                    >
                      <img
                        alt="Lock and Love Women's Removable Hooded Faux Leather Moto Biker Jacket"
                        class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-aob7bp-MuiCardMedia-root"
                        src="https://fakestoreapi.com/img/81XH0e8fefL._AC_UY879_.jpg"
                      />
                      <div
                        class="MuiCardContent-root css-erizva-MuiCardContent-root"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-56rs9p-MuiTypography-root"
                        >
                          Lock and Love Women's Removable Hooded Faux Leather Moto Biker Jacket
                        </p>
                        <p
                          class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-15ejcyd-MuiTypography-root"
                        >
                          Â£
                          29.95
                        </p>
                        <div
                          class="MuiBox-root css-70qvj9"
                        >
                          <p
                            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                          >
                            2.9
                          </p>
                          <span
                            aria-label="3 Stars"
                            class="MuiRating-root MuiRating-sizeMedium Mui-readOnly MuiRating-readOnly css-15befwp-MuiRating-root"
                            role="img"
                          >
                            <span>
                              <span
                                class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                                  data-testid="StarIcon"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                                  />
                                </svg>
                              </span>
                            </span>
                            <span>
                              <span
                                class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                                  data-testid="StarIcon"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                                  />
                                </svg>
                              </span>
                            </span>
                            <span>
                              <span
                                class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                                  data-testid="StarIcon"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                                  />
                                </svg>
                              </span>
                            </span>
                            <span>
                              <span
                                class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                                  data-testid="StarBorderIcon"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                                  />
                                </svg>
                              </span>
                            </span>
                            <span>
                              <span
                                class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                                  data-testid="StarBorderIcon"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </span>
                          <p
                            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                          >
                            340
                             reviews
                          </p>
                        </div>
                      </div>
                    </div>
                  </a>
                </div>
              </div>
            </div>
          </main>
        </div>
      </div>
      <footer
        class="flex justify-around mt-40 p-4 bg-slate-800 text-white"
      >
        <p>
          Copyright Â© 
          2024
           - All right reserved by
           
          E-Shop
        </p>
      </footer>
    </div>
  </body>,
  "container": <div>
    <nav
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1622sjm-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1j0emvn-MuiContainer-root"
      >
        <header
          class="flex justify-between items-center py-4"
        >
          <a
            class="font-semibold text-2xl text-slate-800"
            href="/"
          >
            E-Shop
          </a>
          <nav
            class="hidden sm:block"
          >
            <ul
              class="flex gap-8"
            >
              <li>
                <a
                  class="text-slate-800 font-medium"
                  href="/"
                >
                  All Products
                </a>
              </li>
              <span
                class="MuiBadge-root css-1c32n2y-MuiBadge-root"
              >
                <li>
                  <a
                    class="text-slate-800 font-medium"
                    href="/cart"
                  >
                    Cart
                  </a>
                </li>
                <span
                  class="MuiBadge-badge MuiBadge-standard MuiBadge-invisible MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorPrimary css-y2h6q0-MuiBadge-badge"
                />
              </span>
            </ul>
          </nav>
          <button
            aria-label="open drawer"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium bg-slate-800 css-ddp6wk-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="MenuIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </header>
      </div>
    </nav>
    <div
      class="MuiBox-root css-i9gxme"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1j0emvn-MuiContainer-root"
      >
        <main
          class="pt-[89px]"
        >
          <div
            class="MuiBox-root css-0"
          >
            <h3
              class="MuiTypography-root MuiTypography-h3 css-a5qzwy-MuiTypography-root"
            >
              Products
            </h3>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-spacing-md-3 css-1bzsu2l-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 css-1xucbe2-MuiGrid-root"
              >
                <a
                  href="/product/16"
                >
                  <div
                    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
                  >
                    <img
                      alt="Lock and Love Women's Removable Hooded Faux Leather Moto Biker Jacket"
                      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-aob7bp-MuiCardMedia-root"
                      src="https://fakestoreapi.com/img/81XH0e8fefL._AC_UY879_.jpg"
                    />
                    <div
                      class="MuiCardContent-root css-erizva-MuiCardContent-root"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-56rs9p-MuiTypography-root"
                      >
                        Lock and Love Women's Removable Hooded Faux Leather Moto Biker Jacket
                      </p>
                      <p
                        class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-15ejcyd-MuiTypography-root"
                      >
                        Â£
                        29.95
                      </p>
                      <div
                        class="MuiBox-root css-70qvj9"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                        >
                          2.9
                        </p>
                        <span
                          aria-label="3 Stars"
                          class="MuiRating-root MuiRating-sizeMedium Mui-readOnly MuiRating-readOnly css-15befwp-MuiRating-root"
                          role="img"
                        >
                          <span>
                            <span
                              class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                                data-testid="StarIcon"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                                />
                              </svg>
                            </span>
                          </span>
                          <span>
                            <span
                              class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                                data-testid="StarIcon"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                                />
                              </svg>
                            </span>
                          </span>
                          <span>
                            <span
                              class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                                data-testid="StarIcon"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                                />
                              </svg>
                            </span>
                          </span>
                          <span>
                            <span
                              class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                                data-testid="StarBorderIcon"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                                />
                              </svg>
                            </span>
                          </span>
                          <span>
                            <span
                              class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                                data-testid="StarBorderIcon"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                                />
                              </svg>
                            </span>
                          </span>
                        </span>
                        <p
                          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                        >
                          340
                           reviews
                        </p>
                      </div>
                    </div>
                  </div>
                </a>
              </div>
            </div>
          </div>
        </main>
      </div>
    </div>
    <footer
      class="flex justify-around mt-40 p-4 bg-slate-800 text-white"
    >
      <p>
        Copyright Â© 
        2024
         - All right reserved by
         
        E-Shop
      </p>
    </footer>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
